torch/csrc/jit/tensorexpr/codegen.cpp

```diff
@@ -83,21 +83,134 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
+  size_t size = elementSize(buf->dtype().scalar_type()) * buf->dtype().lanes();
+  for (auto& d : buf->dims()) {
+    if (!d->isConstant()) {
+      return c10::nullopt;
+    }
+    size = size * (*intValue(d));
+  }
+  return size;
+}
+
+using BufRangeInfo =
+    std::unordered_map<BufPtr, std::pair<BufAccessNode, BufAccessNode>>;
+
+std::vector<std::pair<BufPtr, BufPtr>> linearScan(
+    std::unordered_set<BufPtr>& bufs,
+    BufRangeInfo& buf_ranges) {
+  // Sort buffers by the time they appear.
+  std::vector<BufPtr> bufs_sorted(bufs.begin(), bufs.end());
+  auto sorting_function_by_start_time = [&buf_ranges](
+                                            BufPtr b1, BufPtr b2) -> bool {
+    auto start1 = buf_ranges.at(b1).first;
+    auto start2 = buf_ranges.at(b2).first;
+    return std::get<2>(start1).at(0) < std::get<2>(start2).at(0);
+  };
+  std::sort(
+      bufs_sorted.begin(), bufs_sorted.end(), sorting_function_by_start_time);
+
+  auto sorting_function_by_end_time = [&buf_ranges](
+                                          BufPtr b1, BufPtr b2) -> bool {
+    auto end1 = buf_ranges.at(b1).second;
+    auto end2 = buf_ranges.at(b2).second;
+    return std::get<2>(end1).at(0) < std::get<2>(end2).at(0);
+  };
+
+  // Map intermediate buffers to the most recent used memory if any.
+  std::unordered_set<BufPtr> mm;
+  std::list<BufPtr> mm_free;
+  std::unordered_map<BufPtr, BufPtr> b2m;
+  std::vector<std::pair<BufPtr, BufPtr>> b2m_ret;
+
+  std::vector<BufPtr> buf_to_release;
+  for (auto buf : bufs_sorted) {
+    auto start = buf_ranges.at(buf).first;
+    auto end = buf_ranges.at(buf).second;
+
+    // Release memory for buffers whose live range ends before the creation time
+    // of this buf.
+    // TODO: optimize in-place opererations and copy operations
+    buf_to_release.clear();
+    for (auto& mapped : b2m) {
+      auto buf_mapped = mapped.first;
+      auto end_buf_mapped = buf_ranges.at(buf_mapped).second;
+      if (std::get<2>(end_buf_mapped).at(0) < std::get<2>(start).at(0)) {
+        buf_to_release.push_back(buf_mapped);
+      }
+    }
+    std::sort(
+        buf_to_release.begin(),
+        buf_to_release.end(),
+        sorting_function_by_end_time);
+    for (auto& buf_rl : buf_to_release) {
+      mm_free.push_front(b2m[buf_rl]);
+      b2m.erase(buf_rl);
+    }
+
+    // If the buf has dynamic shapes, we'll skip it (i.e., allocate memory for
+    // it, and there are no future reuses on its memory).
+    // TODO: reuse memory for bufs with dynamic shapes
+    if (!bufSize(buf)) {
+      b2m_ret.emplace_back(std::make_pair(buf, buf));
+      continue;
+    }
+
+    bool allocated = false;
+    // Check whether there are free memories that this buf can reuse.
+    for (auto it = mm_free.begin(); it != mm_free.end(); it++) {
+      auto m = *it;
+      TORCH_INTERNAL_ASSERT(bufSize(buf) != 0);
+      TORCH_INTERNAL_ASSERT(bufSize(m) != 0);
+      if (bufSize(m) >= bufSize(buf)) {
```

Comment:
Nit: save bufSize(m) and bufSIze(buf) in variables to avoid recomputing them again.

Reply:
How does this comparison work if `bufSize(m)` returns `c10::nullopt`? Should we exclude such bufs from `mem_up_for_grabs`?